src/loki/search.cc

```diff
@@ -36,7 +38,8 @@ bool is_search_filter_triggered(const DirectedEdge* edge,
   if ((road_class > min_road_class || road_class < max_road_class) ||
       (search_filter.exclude_tunnel_ && edge->tunnel()) ||
       (search_filter.exclude_bridge_ && edge->bridge()) ||
-      (search_filter.exclude_ramp_ && (edge->use() == Use::kRamp))) {
+      (search_filter.exclude_ramp_ && (edge->use() == Use::kRamp)) ||
+      (search_filter.exclude_closures_ && !costing.Filter(edge, tile))) {
```

Comment:
it doesnt make sense for us to rely on the meaning of the costing filter here. for starters, this code path could never be hit if `!costing.Filter` is `true` because it means we would have filtered the edge out in the first place and never checked the search filters for it.

I suggest we just check if you want to exclude closures and if you are a costing that cares about current flow and current flow is 0. ie this:

```suggestion
      (search_filter.exclude_closures_ && (costing.flow_mask() & baldr::kCurrentFlowMask) && tile->IsClosed(edge))) {
```

Reply:
>for starters, this code path could never be hit if !costing.Filter is true because it means we would have filtered the edge out in the first place and never checked the search filters for it

You're right, in that `is_search_filter_triggered` is never called, **IF** `exclude_closures` is `true`. However, if `exclude_closure` is `false`, we internally make `filter_closures: true` which results in `!costing.Filter` returning `false` and we end up invoking the search filter.

Basically  there is never a case where we want `Costing::Filter()` to pass a candidate which might be filtered due to closures,  but then have `is_search_filter_triggered` filter it out. 

The change you suggested won't work if `ignore_closures: true` and `exclude_closures: true`, because the candidates will pass `Costing::Filter` but then fail in `is_search_filter_triggered`. So either we access the state of `ignore_closures` here, or use `DynmaicCost::IsClosed` by making it public (as that already takes `ignore_closures` into account).

I'm leaning towards using the `ignore_closures` value from the costing options pbf. Something like:
```
search_filter.exclude_closures = !costing_options.ignore_closures && location_options.exclude_closures
```
and then use it the way you suggested in `is_search_filter_triggered`. What do you think?

Reply:
i guess im just saying that doing:

```c++
search_filter.exclude_closures_ && !costing.Filter(edge, tile)
```

this code says that if i want to exclude closures and the edge wasnt filtered before filter it now. it not being filtered before is not the same as saying it currently has a closure. there are many reasons an edge can be filtered, closures are not the only ones.

let me think about this a bit more. there simply has to be a way to make this less confusing :smile:

Reply:
> i guess im just saying that it makes no sense to do:
> 
> ```c++
> search_filter.exclude_closures_ && !costing.Filter(edge, tile)
> ```
> 
> this code says that if i want to exclude closures and the edge wasnt filtered before filter it now. it not being filtered before is not the same as saying it currently has a closure. there are many reasons an edge can be filtered, a closure is not the only reason.
> 

Understood. We don't have to use costing filter here. Thats why I suggested accessing `ignore_closures` from the costing options here (or make `DynmaicCost::IsClosed` public). We'll need it to respect its override behavior.